### ISSUE TITLE
Internal Server Error

### ISSUE THREAD

**Issue by fabiostocco** [h2o/h2o#1735]
I'm trying to display pages in php and install wordpress, returns this error:
Internal Server Error

Config:

Linux 18.04 LTS Ionic
PHP7.2

H2o running

h2o.conf

server-name: ""
#user: debian-h2o
user: www-data
access-log: "|rotatelogs -l -f -L /var/log/h2o/access.log -p /usr/share/h2o/compress_logs /var/log/h2o/access.log.%Y-%m-%d 86400"
error-log: "|rotatelogs -l -f -L /var/log/h2o/error.log -p /usr/share/h2o/compress_logs /var/log/h2o/error.log.%Y-%m-%d 86400"
pid-file: /run/h2o.pid

listen:
  port: 80

hosts:
  default:
    paths:
      "/":
        file.dir: /var/www/wordpress

      /server-status:
        status: ON

file.index: [ 'index.php', 'index.html' ]

file.custom-handler:
  extension: .php
  fastcgi.connect:
    port: /run/php-fpm-7.2.sock
    type: unix

send-server-name: OFF
setenv:
  HTTP_PROXY: ""

**Comment by proyb6** [comment]
What is in your access and error log?

**Comment by fabiostocco** [comment]
10.24.88.14 - - [21/Apr/2018:12:32:30 +0000] "GET / HTTP/1.0" 503 21 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36 Edge/16.16299"
10.24.88.14 - - [21/Apr/2018:12:32:30 +0000] "GET /favicon.ico HTTP/1.0" 404 9 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36 Edge/16.16299"
"/var/log/h2o/access.log" 50L, 10288C   


PS: Html files work

**Comment by proyb6** [comment]
I see something missing, perhaps, there is a guide you can try:
https://www.vultr.com/docs/install-the-h2o-web-server-on-centos-7

**Comment by fabiostocco** [comment]
Yes, I understand that it is missing the favicon.ico that is correct ... but that's not the problem!

These instructions of the vultr I have analyzed and verified, what I have realized is that they put:
 redirect:
          internal: YES
          status: 307
          url: /index.php

For me it does not work, h2o does not start, removing everything works!

I also noticed the www.conf and the php-fpm.conf file, I did not see anything that can be done, my files are:

www.conf

[www]
user = www-data
group = www-data
listen = /run/php/php7.2-fpm.sock
listen.owner = www-data
listen.group = www-data
listen.mode = 0666
pm = ondemand
pm.max_children = 5
pm.process_idle_timeout = 10s
pm.max_requests = 2000
chdir = /


php-fpm.conf

[global]
daemonize = yes
emergency_restart_threshold = 2
emergency_restart_interval = 1m
error_log = /var/log/php-fpm-7.2.log
pid = /var/run/php/php-fpm-7.2.pid
process_control_timeout = 10s

error:

root @ h2o: / # curl -I 10.24.88.7
HTTP / 1.1 503 Internal Server Error
Date: Sat, 21 Apr 2018 20:49:48 GMT
Connection: keep-alive
Content-Length: 21
content-type: text / plain; charset = utf-8


Unfortunately I still can not see a solution!

Thank you for your help!

**Comment by proyb6** [comment]
I mean the missing is in the configuration.

I’m assume you have www-data user account and existed in /etc/passwd?

Try removing configure file with “pid-file”

What is in the “sudo netstat -tulpn” when you ran both PHP and H2O?

I would also use “strace”, you can install from your repo, to debug all errors with human readable log that may show you the actual cause, I learnt a file was missing.

https://www.tecmint.com/strace-commands-for-troubleshooting-and-debugging-linux/

**Comment by fabiostocco** [comment]
Yes the user exists, I also tried to use the user "debian-h2o", but it is not the way!
 
I tried to remove the "pid-file" but it did not change at all ..

With your help I could see through the strace command that at the end of debugging there is an error:

write (2, "[/etc/h2o/h2o.conf:9]", 22 [/etc/h2o/h2o.conf:9]) = 22
ANY: 80: Address already in use) = 55 (ANY: 80)
write (2, "\ n", 1
) = 1
exit_group (78) =?
+++ exited with 78 +++

I tried changing the port in "h2o.conf", when I change the h2o starts, I execute the strace command and the warning continues, however with new port that I modified in the file "h2o.conf"

And when I run the "sudo netstat -tulpn" command, it seems to me that it listens perfectly to the ports:

Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address Foreign Address State PID / Program name
tcp 0 0 0.0.0.0:80 0.0.0.0:* LISTEN 6073 / perl
tcp6 0 0 ::: 80 ::: * LISTEN 6073 / perl
tcp6 0 0 ::: 8080 ::: * LISTEN 101 / apache2
udp 0 0 127.0.0.53:53 0.0.0.0:* 76 / systemd-resolved
udp 0 0 10.24.88.7:68 0.0.0.0:* 55 / systemd-networkd
udp6 0 0 fe80 :: 216: 3eff: fe89: 546 ::: * 55 / systemd-networkd

while at php7.2-fpm, everything seems to go perfectly with h2o ...

php7.2-fpm.service - The PHP 7.2 FastCGI Process Manager
    Loaded: loaded (/lib/systemd/system/php7.2-fpm.service; enabled; vendor preset: enabled)
    Active: active (running) since Sat 2018-04-21 20:37:14 UTC; 16 hours ago
      Docs: man: php-fpm7.2 (8)
  Main PID: 2945 (php-fpm7.2)
    Status: "Processes active: 0, idle: 0, Requests: 0, slow: 0, Traffic: 0req / sec"
     Tasks: 1 (limit: 4915)
    CGroup: /system.slice/php7.2-fpm.service
            └─2945 php-fpm: master process (/etc/php/7.2/fpm/php-fpm.conf)

Apr 21 20:37:14 h2o systemd [1]: Starting The PHP 7.2 FastCGI Process Manager ...
Apr 21 20:37:14 h2o systemd [1]: Started The PHP 7.2 FastCGI Process Manager.

H2O server:

h2o.service - Optimized HTTP/1.x, HTTP/2 server
   Loaded: loaded (/lib/systemd/system/h2o.service; enabled; vendor preset: enabled)
   Active: active (running) since Sun 2018-04-22 14:08:41 UTC; 16min ago
     Docs: man:h2o(8)
           file:/usr/share/doc/h2o/index.html
           https://h2o.examp1e.net
 Main PID: 6073 (perl)
    Tasks: 16 (limit: 4915)
   CGroup: /system.slice/h2o.service
           ├─6073 perl -x /usr/share/h2o/start_server --pid-file=/run/h2o.pid --log-file=|rotatelogs -l -f -L /var/log/h2o/error.log -p /usr/share/h2o/compress_logs /va
           ├─6074 rotatelogs -l -f -L /var/log/h2o/error.log -p /usr/share/h2o/compress_logs /var/log/h2o/error.log.%Y-%m-%d 86400
           ├─6076 /usr/bin/h2o -c /etc/h2o/h2o.conf
           ├─6077 /bin/sh -c rotatelogs -l -f -L /var/log/h2o/access.log -p /usr/share/h2o/compress_logs /var/log/h2o/access.log.%Y-%m-%d 86400
           ├─6078 rotatelogs -l -f -L /var/log/h2o/access.log -p /usr/share/h2o/compress_logs /var/log/h2o/access.log.%Y-%m-%d 86400
           └─6079 perl -x /usr/share/h2o/annotate-backtrace-symbols

Apr 22 14:08:41 h2o systemd[1]: Started Optimized HTTP/1.x, HTTP/2 server.
Apr 22 14:08:41 h2o h2o[6073]: start_server (pid:6073) starting now...


so if my understanding is correct, it seems that the error is in listing the door, which in my ignorance seems very strange, because if you insert a page in html, it works perfectly! of course removing the "h2o.conf" file from the php settings

Any way do you visualize?

**Comment by proyb6** [comment]
From my fast reading, 6073/Perl is using network port 80, so H2O could not started on the same port 80.

Perhaps, you might want to terminate Perl if nothing critical is running and rerun H2O.

**Comment by proyb6** [comment]
Sorry, I’m not too familiar with PHP7 since I had done PHP5 years ago, currently working with Nodejs and H2O is less of a problem for my use case, perhaps others can help.

**Comment by fabiostocco** [comment]
Thank you very much for your attention, success!

**Comment by mckaygerhard** [comment]
this sucks.. the problem is the h2o insane level of security.. and is clear h2o cannot access the socket of the php running service cos php running service saids must be under www-data group but h2o does not understand that!
